src/Native/Unix/System.Native/pal_io.cpp

```diff
@@ -360,6 +363,28 @@ extern "C" int32_t SystemNative_ReadDirR(DIR* dir, void* buffer, int32_t bufferS
 
     // 0 returned with non-null result (guaranteed to be set to entry arg) -> success
     assert(result == entry);
+#else
+    errno = 0;
+    result = readdir(dir);
+
+    //  kernel set errno -> failure
```

Comment:
For some reason, I was thinking that readdir() had result passed to it and could do whatever it wanted with it internally.